File: apps/cli/internal/heal/tools/registry.go
package tools

import (
	"context"
	"encoding/json"
	"fmt"

	"github.com/anthropics/anthropic-sdk-go"
)

// Registry holds all registered tools and handles dispatch.
type Registry struct {
	tools   map[string]Tool
	toolCtx *Context
}

// NewRegistry creates a new tool registry with the given context.
func NewRegistry(ctx *Context) *Registry {
	return &Registry{
		tools:   make(map[string]Tool),
		toolCtx: ctx,
	}
}

// Register adds a tool to the registry.
func (r *Registry) Register(tool Tool) {
	r.tools[tool.Name()] = tool
}

// Get returns a tool by name, or nil if not found.
func (r *Registry) Get(name string) Tool {
	return r.tools[name]
}

// Dispatch executes a tool by name with the given input.
// Returns a Result that should be sent back to Claude.
func (r *Registry) Dispatch(ctx context.Context, name string, input json.RawMessage) Result {
	tool := r.tools[name]
	if tool == nil {
		return ErrorResult(fmt.Sprintf("unknown tool: %s", name))
	}

	result, err := tool.Execute(ctx, input)
	if err != nil {
		// Only return Go errors for fatal issues (context cancelled, etc.)
		// Tool execution errors should be in the Result
		return ErrorResult(fmt.Sprintf("tool execution failed: %v", err))
	}

	return result
}

// ToAnthropicTools converts registered tools to Anthropic SDK format.
func (r *Registry) ToAnthropicTools() []anthropic.ToolUnionParam {
	tools := make([]anthropic.ToolUnionParam, 0, len(r.tools))

	for _, tool := range r.tools {
		schema := tool.InputSchema()

		// Extract required fields from schema if present
		var required []string
		if req, ok := schema["required"].([]string); ok {
			required = req
		}

		// Extract properties from schema
		var properties any
		if props, ok := schema["properties"]; ok {
			properties = props
		} else {
			properties = schema
		}

		toolParam := anthropic.ToolParam{
			Name:        tool.Name(),
			Description: anthropic.String(tool.Description()),
			InputSchema: anthropic.ToolInputSchemaParam{
				Properties: properties,
				Required:   required,
			},
		}
		tools = append(tools, anthropic.ToolUnionParam{OfTool: &toolParam})
	}

	return tools
}

// Context returns the tool execution context.
func (r *Registry) Context() *Context {
	return r.toolCtx
}

// Names returns the list of registered tool names.
func (r *Registry) Names() []string {
	names := make([]string, 0, len(r.tools))
	for name := range r.tools {
		names = append(names, name)
	}
	return names
}

File: packages/healing/src/tools/registry.ts
import type { Tool as AnthropicTool } from "@anthropic-ai/sdk/resources/messages.js";
import type { ToolContext } from "./context.js";
import type { Tool, ToolResult } from "./types.js";
import { errorResult } from "./types.js";

/**
 * Registry for managing and dispatching tools.
 */
export class ToolRegistry {
  private readonly tools: Map<string, Tool> = new Map();
  private readonly ctx: ToolContext;
  private cachedAnthropicTools: AnthropicTool[] | null = null;

  constructor(ctx: ToolContext) {
    this.ctx = ctx;
  }

  /**
   * Registers a tool with the registry.
   */
  register = (tool: Tool): void => {
    this.tools.set(tool.name, tool);
    this.cachedAnthropicTools = null;
  };

  /**
   * Registers multiple tools at once.
   */
  registerAll = (tools: Tool[]): void => {
    for (const tool of tools) {
      this.register(tool);
    }
  };

  /**
   * Gets a tool by name.
   */
  get = (name: string): Tool | undefined => this.tools.get(name);

  /**
   * Checks if a tool exists.
   */
  has = (name: string): boolean => this.tools.has(name);

  /**
   * Dispatches a tool call by name with the given input.
   */
  dispatch = async (name: string, input: unknown): Promise<ToolResult> => {
    const tool = this.tools.get(name);
    if (!tool) {
      return errorResult(`unknown tool: ${name}`);
    }

    try {
      return await tool.execute(this.ctx, input);
    } catch (err) {
      const message = err instanceof Error ? err.message : String(err);
      return errorResult(`tool execution failed: ${message}`);
    }
  };

  /**
   * Converts registered tools to Anthropic SDK format.
   * Results are cached until a new tool is registered.
   */
  toAnthropicTools = (): AnthropicTool[] => {
    if (this.cachedAnthropicTools) {
      return this.cachedAnthropicTools;
    }

    this.cachedAnthropicTools = Array.from(this.tools.values()).map(
      (tool): AnthropicTool => ({
        name: tool.name,
        description: tool.description,
        input_schema: {
          type: "object",
          ...tool.inputSchema,
        },
      })
    );

    return this.cachedAnthropicTools;
  };

  /**
   * Returns the number of registered tools.
   */
  get size(): number {
    return this.tools.size;
  }

  /**
   * Returns all tool names.
   */
  get names(): string[] {
    return Array.from(this.tools.keys());
  }
}

/**
 * Creates a new tool registry with the given context.
 */
export const createToolRegistry = (ctx: ToolContext): ToolRegistry =>
  new ToolRegistry(ctx);
